### PR TITLE
feat(WebSocket): AND-1325 Connection events and auto retry websocket

### PR DESCRIPTION
**Pull request by westonal**

### CODE REVIEW

File: common/network/src/main/kotlin/com/blockchain/network/websocket/OkHttpWebSocket.kt [hunk 1]
```diff
@@ -44,23 +64,32 @@ class OkHttpWebSocket(
...
         override fun onMessage(webSocket: okhttp3.WebSocket, message: String) {
+            super.onMessage(webSocket, message)
             subject.onNext(message)
-            listener.onMessage(message)
+            listener?.onMessage(message)
+        }
+
+        override fun onFailure(webSocket: okhttp3.WebSocket, t: Throwable, response: Response?) {
+            super.onFailure(webSocket, t, response)
+            connectionEventsSubject.onNext(ConnectionEvent.Failure(t))
```
Comment (westonal) [inline common/network/src/main/kotlin/com/blockchain/network/websocket/OkHttpWebSocket.kt]: I failed to test drive this :( couldn't get the mock server to trigger it.

File: common/network/src/main/kotlin/com/blockchain/network/websocket/RetryingConnection.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    override fun close() {
+        connections.clear()
+        inner.close()
+    }
+}
+
+private fun AtomicLong.double() {
+    while (!get().let { compareAndSet(it, it * 2) }) {
+    }
+}
```
Comment (westonal) [inline common/network/src/main/kotlin/com/blockchain/network/websocket/RetryingConnection.kt]: Oops, will remove

File: common/network/src/main/kotlin/com/blockchain/network/websocket/RetryingConnection.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        connectionEvents
+            .throttleWithTimeout(timeoutTimes[timeoutIndex], TimeUnit.MILLISECONDS)
+            .subscribe {
+                when (it) {
+                    is ConnectionEvent.Failure -> {
+                        resetEvents(Math.min(timeoutIndex + 1, timeoutTimes.size - 1))
+                        inner.open()
+                    }
+                    ConnectionEvent.ClientDisconnect -> inner.open()
+                    ConnectionEvent.Connected -> {
```
Comment (ditn) [inline common/network/src/main/kotlin/com/blockchain/network/websocket/RetryingConnection.kt]: Could remove the brackets here, just stylistic.

Comment (westonal) [reply common/network/src/main/kotlin/com/blockchain/network/websocket/RetryingConnection.kt]: sure, left over from refactor

**Pull request merged**